File: zig-generator/src/emitters/store_to_load.zig
const std = @import("std");
const Emitter = @import("../emitter.zig").Emitter;
const GeneratorState = @import("../state.zig").GeneratorState;

pub const StoreToLoadEmitter = struct {
    allocator: std.mem.Allocator,

    pub fn emitter(self: *StoreToLoadEmitter) Emitter {
        return .{ .ptr = self, .vtable = &.{ .try_emit = tryEmit } };
    }


  fn tryEmit(ptr: *anyopaque, state: *GeneratorState) ?[]u8 {
      const self: *StoreToLoadEmitter = @ptrCast(@alignCast(ptr));
      const scratch_id = state.getScratchRegionId() orelse return null;
      const scratch = state.getRegion(scratch_id);
      if (scratch.watermark >= scratch.size) return null;
      const src    = state.pickInitializedGpr() orelse return null;
      const dst    = state.pickWritableGpr()    orelse return null;
      const offset = scratch.watermark;
      state.hitCoverage(.StoreToLoadFwd);
      state.advanceWatermark(scratch_id, 8);
      state.markScalar(dst);
      return std.fmt.allocPrint(self.allocator,
          "  STR x{d}, [x22, #{d}]\n  LDR x{d}, [x22, #{d}]",
          .{ src, offset, dst, offset },
      ) catch null;
  }
};

File: rust-generator/src/emitters/store_to_load.rs
use crate::coverage::CovPoint;
use crate::emitter::Emitter;
use crate::state::GeneratorState;

pub struct StoreToLoadEmitter;

impl Emitter for StoreToLoadEmitter {
    fn try_emit(&mut self, state: &mut GeneratorState) -> Option<String> {
        let scratch_id = state.get_scratch_region_id()?;
        let (watermark, size) = {
            let s = state.get_region(scratch_id);
            (s.watermark, s.size)
        };
        if watermark >= size { return None; }
        let src = state.pick_initialized_gpr()?;
        let dst = state.pick_writable_gpr()?;
        state.hit_coverage(CovPoint::StoreToLoadFwd);
        state.advance_watermark(scratch_id, 8);
        state.mark_scalar(dst);
        Some(format!(
            "    STR x{}, [x22, #{}]\n    LDR x{}, [x22, #{}]",
            src, watermark, dst, watermark
        ))
    }
}
